v0.7/speech_recognition/rnnt/pytorch_SUT.py

```diff
@@ -0,0 +1,108 @@
+# Copyright (c) 2020, Cerebras Systems, Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#           http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import sys
+import os
+sys.path.insert(0, os.path.join(os.getcwd(), "pytorch"))
+
+import array
+import torch
+import numpy as np
+import toml
+import mlperf_loadgen as lg
+
+from QSL import AudioQSL
+from decoders import ScriptGreedyDecoder
+from helpers import add_blank_label
+from preprocessing import AudioPreprocessing
+from model_separable_rnnt import RNNT
+
+
+def load_and_migrate_checkpoint(ckpt_path):
+    checkpoint = torch.load(ckpt_path, map_location="cpu")
+    migrated_state_dict = {}
+    for key, value in checkpoint['state_dict'].items():
+        key = key.replace("joint_net", "joint.net")
+        migrated_state_dict[key] = value
+    del migrated_state_dict["audio_preprocessor.featurizer.fb"]
+    del migrated_state_dict["audio_preprocessor.featurizer.window"]
+    return migrated_state_dict
+
+
+class PytorchSUT:
+    def __init__(self, config_toml, checkpoint_path, dataset_dir,
+                 manifest_filepath, perf_count):
+        config = toml.load(config_toml)
+
+        dataset_vocab = config['labels']['labels']
+        rnnt_vocab = add_blank_label(dataset_vocab)
+        featurizer_config = config['input_eval']
+
+        self.sut = lg.ConstructSUT(self.issue_queries, self.flush_queries,
+                                   self.process_latencies)
+        # TODO: What to do about perf_count?
+        self.qsl = AudioQSL(dataset_dir, manifest_filepath,
+                            dataset_vocab, featurizer_config["sample_rate"])
+
+        self.audio_preprocessor = AudioPreprocessing(**featurizer_config)
+
+        model = RNNT(
+            feature_config=featurizer_config,
+            rnnt=config['rnnt'],
+            num_classes=len(rnnt_vocab)
+        )
+        model.load_state_dict(load_and_migrate_checkpoint(checkpoint_path),
+                              strict=True)
+        model.eval()
+        model = torch.jit.script(model)
+        self.greedy_decoder = ScriptGreedyDecoder(len(rnnt_vocab) - 1, model)
+
+    def issue_queries(self, query_samples):
+        for query_sample in query_samples:
```

Comment:
I simply do a for-loop through the data, regardless of what mode is active (server, offline, multistream, etc.)

I see that v0.7 BERT does this, but it looks like v0.5 GNMT is more sophisticated, since it does batching when not doing single-stream (among other things). I was a little confused based on this rule: https://github.com/mlperf/inference_policies/blob/master/inference_rules.adoc#26-benchmark-detection-is-not-allowed

I'm not sure what it means to detect a benchmark. Does that mean the scenario (offline, server, single stream, multistream)? @petermattson 

What all is expected here on my side?

Reply:
benchmark detection means having special HW settings when running benchmarks but not intended for general use. I don't think it is applicable to your SW here.

Reply:
This is fine. This is always running samples one sample after another, so it's not "detecting" any scenarios :)